doc/developer/design/20210408_timelines.md

```diff
@@ -0,0 +1,101 @@
+We should prevent this poor UX from happening to users.
+
+In addition to an infinite delay, there are cases where, even if the numbers were the same, they are meaningless when related to each other.
+For example, two separate Debezium consistency topics.
+Although they both start counting at 1 and thus have relatively close numbers, there's no meaning in the transaction counter, and joining data on that will not produce results that are meaningful.
+We should inform users when they attempt to join data from differing timelines that they have not asked us a useful question.
+
+The third concern is linearizability.
+We would like to provide that property when users issue SELECT queries.
+If we can distinguish the meaning of timestamps across data, we can provide linearizability across sources correctly.
+(This will be described with more detail in a later doc.)
+
+A solution to this problem is to inform the system what the timestamp number means for all data, and allow the system to reject queries that request data from multiple meanings.
+We introduce a thing called a timeline which is the meaning of the timestamp number.
+At the points in the system at which it is possible to join different sources together, we will return an error to the user if there is more than one timeline present.
+There are two places where this can occur: `CREATE VIEW` and `SELECT`.
+These have been taught to check that their expression depends at most one timeline.
+Other operations that can create dataflow indexes (TAIL, CREATE SOURCE, CREATE SINK, CREATE INDEX) operate on existing views, and are thus unable to create dataflows from multiple timelines.
+
+This is implemented by adding a `Timeline` enum, and requiring sources and tables to provide their `Timeline` to the Coordinator.
+When a VIEW or SELECT is planned, its MirRelationExpr is crawled to discover all timelines.
+
+Variants of `Timeline`:
+
+- `EpochMilliseconds` for anything producing timestamps near wall-clock now.
+- `Counter(struct)` for Debezium consistency topics, with a struct to track broker/topic to tell apart different sources.
+- `External(String)` for CDCv2 sources with the source's name attached, making a CDCv2 source only joinable with itself.
+- `User(String)` for user-defined timelines by a name chosen by the user.
+
+All sources can specify `WITH (timeline='some_name')` to choose the `User` variant.
+This will group any set of sources together if users need to join CDCv2 and realtime sources, or multiple CDCv2 sources.
+
+By default, a CDCv2 source is joinable only with itself.
+We do not allow them to join with even other CDCv2 sources because of side effects in linearizability and time domains, which might like to advance to the most recent time.
+CDCv2 sources have a special `WITH (system_timeline=true)` syntax to put that source into the EpochMilliseconds timeline.
```

Comment:
Does it make sense to call this "epoch_ms_timeline" to be more clear? I liked the use of `EpochMilliseconds` up above.

Reply:
Both names have their appeal. system_timeline I thought would match our intended documentation (change out realtime for system time) better which is why I chose that? I don't care either way.

Reply:
After a sleep I think epoch_ms_timeline is better because system feels very dependent on the machine a thing is running on, and epoch_ms feels more descriptive about the exact data we expect.